### PR TITLE
Remove VLA from flex_trit/tryte conversion functions

### LINKED ISSUE #360: Remove VLA from flex_trit/tryte conversion functions
The flex_trit/tryte conversion functions make use of variable length arrays, such as:
```trit_t trits[num_trits];```
and
```trit_t trits[to_len];```
Develop appropriate algorithm to avoid VLA

### PR DESCRIPTION
**Pull request by ifullgaz**
Remove VLA from flex_trit/tryte conversion functions (fixes #360)
Depends on PR #346 and PR #362 

# Test Plan:
Usual test plan

### CODE REVIEW

**Comment by th0br0** [comment]
Could you also make `num_flex_trits_for_trits` a #define macro as part of this PR?

**Comment by ifullgaz** [comment]
You mean instead of inline?

File: common/trinary/flex_trit.c [hunk 1]
```diff
@@ -166,12 +166,42 @@ size_t flex_trits_to_trytes(tryte_t *trytes, size_t to_len,
...
 #elif defined(FLEX_TRIT_ENCODING_3_TRITS_PER_BYTE)
-  memcpy(trytes, flex_trits, num_flex_trits_for_trits(num_trits));
+  uint8_t residual = num_trits % 3U;
+  size_t num_bytes = num_flex_trits_for_trits(num_trits);
+  // There is a risk of noise after the last trit so we need to clean up
+  if (residual) {
+    trit_t trits[3];
+    num_bytes--;
+    trytes_to_trits((tryte_t *)(flex_trits + num_bytes), trits, 1);
+    trits_to_trytes(trits, (tryte_t *)(trytes + num_bytes), residual);
```
Comment (tsvisabo) [inline common/trinary/flex_trit.c:176]: I think it's better here to only `assert (num_trits % 3U == 0)`

File: common/trinary/flex_trit.c [hunk 2]
```diff
@@ -184,13 +184,13 @@ size_t flex_trits_to_trytes(tryte_t *trytes, size_t to_len,
   };
   union _shifter shifter = {0};
   size_t offset = 0;
-  size_t max_trits, trits_for_tryte;
-  for (int i = 0, j = 0; num_trits || offset; j++) {
+  size_t max_trits, trits_for_tryte, n_trits = num_trits;
+  for (int i = 0, j = 0; n_trits || offset; j++) {
```
Comment (tsvisabo) [inline common/trinary/flex_trit.c]: can we rename n_trits ?

File: common/trinary/flex_trit.c [hunk 3]
```diff
@@ -166,12 +166,42 @@ size_t flex_trits_to_trytes(tryte_t *trytes, size_t to_len,
...
+      max_trits = MIN(NUM_TRITS_PER_FLEX_TRIT, num_trits);
+      flex_trits_to_trits(shifter.trits + offset, max_trits, &flex_trits[i++],
+                          max_trits, max_trits);
+      num_trits -= max_trits;
+      offset += max_trits;
+    }
+    trits_for_tryte = MIN(3, offset);
+    trits_to_trytes(shifter.trits, &trytes[j], trits_for_tryte);
+#if BYTE_ORDER == LITTLE_ENDIAN
+    shifter.val = shifter.val >> 24;
```
Comment (tsvisabo) [inline common/trinary/flex_trit.c:199]: where does `shifter.val` being used?

File: common/trinary/flex_trit.c [hunk 4]
```diff
@@ -190,10 +220,34 @@ size_t flex_trits_from_trytes(flex_trit_t *to_flex_trits, size_t to_len,
...
+        i++;
+        trytes_to_trits(&trytes[i], shifter.trits + offset, 1);
+        num_trytes -= 1;
+        offset += 3;
+      }
+    }
+    num_trits = MIN(NUM_TRITS_PER_FLEX_TRIT, offset);
+    flex_trits_from_trits(to_flex_trits + j, num_trits, shifter.trits, num_trits, num_trits);
+#if BYTE_ORDER == LITTLE_ENDIAN
+    shifter.val = shifter.val >> (num_trits << 3);
```
Comment (tsvisabo) [inline common/trinary/flex_trit.c:246]: where is it being used?

File: common/trinary/flex_trit.c [hunk 5]
```diff
@@ -166,12 +166,42 @@ size_t flex_trits_to_trytes(tryte_t *trytes, size_t to_len,
...
+  memcpy(trytes, flex_trits, num_bytes);
 #elif defined(FLEX_TRIT_ENCODING_4_TRITS_PER_BYTE) || \
-    defined(FLEX_TRIT_ENCODING_5_TRITS_PER_BYTE)
-  trit_t trits[num_trits];
-  flex_trits_to_trits(trits, num_trits, flex_trits, len, num_trits);
-  trits_to_trytes((trit_t *)trits, trytes, num_trits);
+      defined(FLEX_TRIT_ENCODING_5_TRITS_PER_BYTE)
+  union _shifter {
+    uint64_t val;
+    trit_t trits[8];
```
Comment (tsvisabo) [inline common/trinary/flex_trit.c:183]: 8? not 7?

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: common/trinary/flex_trit.c (hunk 1 above)
Comment (ifullgaz) [reply common/trinary/flex_trit.c:176]: If num_trits %3U == 0 then this code is not ran. Anywhere else we accept num_trits as a parameter, there is no such constraint. At least, this way, all possibilities are taken into account.

File: common/trinary/flex_trit.c (hunk 2 above)
Comment (ifullgaz) [reply common/trinary/flex_trit.c]: Renamed to trits_counter

File: common/trinary/flex_trit.c (hunk 3 above)
Comment (ifullgaz) [reply common/trinary/flex_trit.c:199]: It is used to shift the bytes of `shifter.trits`. In this case, 24 bits = 3 bytes

File: common/trinary/flex_trit.c (hunk 4 above)
Comment (ifullgaz) [reply common/trinary/flex_trit.c:246]: It is used to shift the bytes of `shifter.trits`. In this case, `num_trits` bytes

File: common/trinary/flex_trit.c (hunk 5 above)
Comment (ifullgaz) [reply common/trinary/flex_trit.c:183]: 8 bytes for a 64 bits value

File: common/trinary/flex_trit.c (hunk 4 above)
Comment (tsvisabo) [reply common/trinary/flex_trit.c:246]: my bad

File: common/trinary/flex_trit.c (hunk 1 above)
Comment (tsvisabo) [reply common/trinary/flex_trit.c:176]: ok

**Review by tsvisabo — APPROVED** [review]

**Pull request merged**